main/src/ca/uwaterloo/flix/language/phase/Safety.scala

```diff
@@ -492,6 +493,25 @@ object Safety {
       }
   }
 
+  /** Returns all concrete types for all associated types in `tpe0`. */
+  private def unpackAssocType(tpe0: Type)(implicit root: Root): Type = tpe0 match {
+    case Type.Var(_, _) => tpe0
+    case Type.Cst(_, _) => tpe0
+    case Type.Apply(tpe1, tpe2, loc) => Type.Apply(unpackAssocType(tpe1), unpackAssocType(tpe2), loc)
+    case Type.Alias(symUse, args, tpe, loc) => Type.Alias(symUse, args.map(unpackAssocType), unpackAssocType(tpe), loc)
+    case Type.AssocType(SymUse.AssocTypeSymUse(sym, _), arg, _, _) =>
+      // Recurse on arg
+      val t = unpackAssocType(arg)
+
+      // Find instance of `sym.trt` with the given type arg (guaranteed to exist)
+      val inst = root.instances(sym.trt).find(i => i.tpe == t).get
```

Comment:
@mlutze Is this guaranteed to exist?

Reply:
We need to know if the signature of the function we are looking at has a `where` clause. If it does, then an associated type applied to a variable can be resolved to something.

e.g. `where Elm[a] ~ String`

But most of the time if there's no constraint like this, it should be an error.